src/main/java/com/minecolonies/structures/helpers/Structure.java

```diff
@@ -527,35 +539,63 @@ public void setPlacementSettings(final PlacementSettings settings)
      */
     public void renderStructure(@NotNull final BlockPos startingPos, @NotNull final World clientWorld, @NotNull final EntityPlayer player, final float partialTicks)
     {
-        final Template.BlockInfo[] blockList = this.getBlockInfoWithSettings(this.settings);
         final Entity[] entityList = this.getEntityInfoWithSettings(clientWorld, startingPos, this.settings);
 
-        for (final Template.BlockInfo aBlockList : blockList)
+        if(!lastStartingPos.equals(startingPos))
         {
-            Block block = aBlockList.blockState.getBlock();
-            IBlockState iblockstate = aBlockList.blockState;
+            modelList.clear();
+        }
 
-            if (block == ModBlocks.blockSubstitution)
-            {
-                continue;
-            }
+        if (modelList.isEmpty())
+        {
+            lastStartingPos = startingPos;
+            final Template.BlockInfo[] blockList = this.getBlockInfoWithSettings(this.settings);
+
+            final FakeWorld fakeWorld = new FakeWorld(null, clientWorld.getSaveHandler(), clientWorld.getWorldInfo(), clientWorld.provider, clientWorld.profiler, true, null, true);
 
-            if (block == ModBlocks.blockSolidSubstitution)
+            for (final Template.BlockInfo aBlockList : blockList)
             {
-                iblockstate = BlockUtils.getSubstitutionBlockAtWorld(clientWorld, startingPos);
-                block = iblockstate.getBlock();
+                final IBlockState iblockstate = aBlockList.blockState;
+                fakeWorld.setBlockState(aBlockList.pos, iblockstate);
+                final Block block = iblockstate.getBlock();
+                TileEntity tileentity = null;
+                if (block.hasTileEntity(aBlockList.blockState) && aBlockList.tileentityData != null)
+                {
+                    tileentity = block.createTileEntity(clientWorld, iblockstate);
+                    tileentity.readFromNBT(aBlockList.tileentityData);
+                }
+                fakeWorld.setTileEntity(aBlockList.pos, tileentity);
             }
 
-            final BlockPos blockpos = aBlockList.pos.add(startingPos);
-            final IBlockState iBlockExtendedState = block.getExtendedState(iblockstate, clientWorld, blockpos);
-            final IBakedModel ibakedmodel = Minecraft.getMinecraft().getBlockRendererDispatcher().getModelForState(iblockstate);
-            TileEntity tileentity = null;
-            if (block.hasTileEntity(iblockstate) && aBlockList.tileentityData != null)
+            for (final Template.BlockInfo aBlockList : blockList)
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Combine this loop with the one that starts on line 556. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS3047)